src/test/shell/bazel/android/android_integration_test.sh

```diff
@@ -464,4 +464,31 @@ EOF
   fi
 }
 
+function test_lots_of_assets {
+  write_hello_android_files
+  setup_android_sdk_support
+
+  cat > java/com/example/hello/BUILD <<'EOF'
+LONG_PATH = "a" * 200
+RANGE = [str(i) for i in range(5000)]
+
+android_binary(
+    name = "hello",
+    manifest = "AndroidManifest.xml",
+    srcs = glob(["*.java"]),
+    resource_files = glob(["res/**"]),
+    assets_dir = LONG_PATH,
+    assets = [":file" + i for i in RANGE],
+)
```

Comment:
This is a good test, but doesn't necessarily test for the intent of this PR, de-duplication of assets. Any idea where the duplication is coming from?

Reply:
> Any idea where the duplication is coming from?

I kinda don't understand the question. Doesn't the diff in AndroidAssets.java here https://github.com/bazelbuild/bazel/pull/17012/files#diff-6659be375abd4d0e15f9a5c525f8be296d586529f8bb8cc45a35f81c3d2edf9aR82 answer that question? That function collects dirs for every asset and stuffs those dirs into a list, thus the same dir can occur multiple times on the command line later, when it calls aapt.

> This is a good test, but doesn't necessarily test for the intent of this PR, de-duplication of assets.

I see what you are saying, it should be a test with two assets in the same dir, and a check that aapt was called with just one of those. I'll see what I can do.

Reply:
> That function collects dirs for every asset and stuffs those dirs into a list, thus the same dir can occur multiple times on the command line later, when it calls aapt.

So this describes the solution to the problem, but I suspect that the problem itself originates from somewhere else in the codebase, right? I'm wondering if it's possible to identify and fix the problem itself, rather than ameliorate the symptom.

Reply:
Oh, actually I just understood what you were saying. `assetsRoots`, when expressed as a List, does not care about duplicate assets when it is appended to: https://github.com/bazelbuild/bazel/pull/17012/files#diff-6659be375abd4d0e15f9a5c525f8be296d586529f8bb8cc45a35f81c3d2edf9aR91

Refactoring `assetRoots` to a Set ensures asset uniqueness by construction. Do you think it's worth adding a comment to explain this rationale?